src/common/ApiResponse/ApiEventsContext.tsx

```diff
@@ -1,27 +1,86 @@
 "use client";
 
-import { PropsWithChildren, createContext, useContext, useState } from "react";
+import {
+  Dispatch,
+  PropsWithChildren,
+  createContext,
+  useContext,
+  useReducer,
+} from "react";
 import { ImmutableMap } from "map-immute";
 
 import { ApiEvent } from "./types";
 
 export type ApiEventsMap = ImmutableMap<string, ApiEvent>;
+export type InitialApiEventsState = Iterable<readonly [string, ApiEvent]>;
+
+export type ApiEventAction =
+  | {
+      type: "set";
+      event: ApiEvent;
+    }
+  | {
+      type: "delete";
+      id: string;
+    }
+  | {
+      type: "clear";
+    };
+
+export function apiEventsReducer(
+  state: ApiEventsMap,
+  action: ApiEventAction
+): ApiEventsMap {
+  switch (action.type) {
+    case "set":
+      return state.set(action.event.id, action.event);
+    case "delete":
+      const result = state.safeDelete(action.id);
+      if (!result) return state;
+      return result;
+    case "clear":
+      return state.safeClear();
+    default:
+      return state;
+  }
+}
```

Comment:
Really liked the ergonomics of using my ImmutableMap package, I was able to build a neat reducer on top of the `ImmutableMap` methods.